### PR TITLE
[Edge] support edge

### PR DESCRIPTION
**Pull request by anychhoice**
- If no client_id is found in the data, publish it to all connected clients.
- Add NNS_EDGE_FLAG_PUBSUB.

Signed-off-by: Yechan Choi <yechan9.choi@samsung.com>

### CODE REVIEW

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c [hunk 1]
```diff
@@ -1368,17 +1391,26 @@ nns_edge_publish (nns_edge_h edge_h, nns_edge_data_h data_h)
...
+        *conn_datas = (nns_edge_conn_data_s * ) malloc (sizeof(nns_edge_conn_data_s *) * length);
+        conn_datas[0] = _nns_edge_get_connection (eh, client_id);
+      } else {
+        _nns_edge_get_all_connections(eh, &conn_datas, &length);
       }
 
-      ret = _nns_edge_transfer_data (conn_data->sink_conn, data_h, client_id);
-      if (ret != NNS_EDGE_ERROR_NONE)
-        nns_edge_loge ("Failed to send edge data.");
+      for (i = 0; i < length; i++) {
```
Comment (jaeyun-jung) [inline src/libnnstreamer-edge/nnstreamer-edge-internal.c:1401]: not enough time to you :)
recommend to use g_hash_table_get_keys_as_array()
```
int
nns_edge_publish (nns_edge_h edge_h, nns_edge_data_h data_h)
{
  nns_edge_handle_s *eh;
  nns_edge_conn_data_s *conn_data;
  int64_t *client_ids;
  guint i, length;
  char *val;
  int ret = NNS_EDGE_ERROR_NONE;

  eh = (nns_edge_handle_s *) edge_h;
  if (!eh) {
    nns_edge_loge ("Invalid param, given edge handle is null.");
    return NNS_EDGE_ERROR_INVALID_PARAMETER;
  }

  if (nns_edge_data_is_valid (data_h) != NNS_EDGE_ERROR_NONE) {
    nns_edge_loge ("Invalid param, given edge data is invalid.");
    return NNS_EDGE_ERROR_INVALID_PARAMETER;
  }

  nns_edge_lock (eh);

  if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
    nns_edge_loge ("Invalid param, given edge handle is invalid.");
    nns_edge_unlock (eh);
    return NNS_EDGE_ERROR_INVALID_PARAMETER;
  }

  ret = nns_edge_data_get_info (data_h, "client_id", &val);
  if (ret != NNS_EDGE_ERROR_NONE) {
    nns_edge_logd
        ("Cannot find client ID in edge data. Publish to all connected clients.");
    client_ids = (int64_t *) g_hash_table_get_keys_as_array (eh->conn_table,
        &length);
  } else {
    length = 1U;

    client_ids = (int64_t *) malloc (sizeof (int64_t));
    if (client_ids)
      *client_ids = strtoll (val, NULL, 10);

    SAFE_FREE (val);
  }

  if (!client_ids) {
    nns_edge_loge ("Failed to get client ID, cannot send edge data.");
    goto done;
  }

  /** @todo update code for each connect type */
  switch (eh->connect_type) {
    case NNS_EDGE_CONNECT_TYPE_TCP:
    case NNS_EDGE_CONNECT_TYPE_HYBRID:
      for (i = 0; i < length; i++) {
        conn_data = _nns_edge_get_connection (eh, client_ids[i]);
        if (!conn_data) {
          nns_edge_loge
              ("Cannot find connection, invalid client ID or connection closed.");
          ret = NNS_EDGE_ERROR_INVALID_PARAMETER;
          break;
        }

        ret = _nns_edge_transfer_data (conn_data->sink_conn, data_h,
            client_ids[i]);
        if (ret != NNS_EDGE_ERROR_NONE) {
          nns_edge_loge ("Failed to send edge data.");
          goto done;
        }
      }
      break;
    default:
      break;
  }

done:
  SAFE_FREE (client_ids);
  nns_edge_unlock (eh);
  return ret;
}
```

Comment (anychhoice) [reply src/libnnstreamer-edge/nnstreamer-edge-internal.c:1411]: Updated.
Thank you :D

File: include/nnstreamer-edge.h [hunk 2]
```diff
@@ -78,8 +78,9 @@ typedef enum {
   NNS_EDGE_FLAG_RECV = (1 << 0),
   NNS_EDGE_FLAG_SEND = (1 << 1),
   NNS_EDGE_FLAG_SERVER = (1 << 2),
+  NNS_EDGE_FLAG_PUBSUB = (1 << 3),
```
Comment (myungjoo) [inline include/nnstreamer-edge.h:81]: What does this do?
(NNS_EDGE_FLAG_PUBSUB? not NNS_EDGE_FLAG_PUB, NNS_EDGE_FLAG_SUB?)

Comment (gichan-jang) [reply include/nnstreamer-edge.h:81]: It is to distinguish between query and pub/sub.
How about these to make it clear?
```
  NNS_EDGE_FLAG_QUERY_CLIENT = (1 << 0),
  NNS_EDGE_FLAG_QUERY_SERVER = (1 << 1),
  NNS_EDGE_FLAG_PUB = (1 << 2),
  NNS_EDGE_FLAG_SUB = (1 << 3),
```
+@jaeyun-jung

Comment (myungjoo) [reply include/nnstreamer-edge.h:81]: > It is to distinguish between query and pub/sub. How about these to make it clear?
> 
> ```
>   NNS_EDGE_FLAG_QUERY_CLIENT = (1 << 0),
>   NNS_EDGE_FLAG_QUERY_SERVER = (1 << 1),
>   NNS_EDGE_FLAG_PUB = (1 << 2),
>   NNS_EDGE_FLAG_SUB = (1 << 3),
> ```
> 
> +@jaeyun-jung

Yes, ```{ QUERY_CLIENT, QUERY_SERVER, PUB, SUB }``` is intuitive and clear.

Comment (myungjoo) [reply include/nnstreamer-edge.h:81]: And be aware that ```nns_edge_flag_e``` is NEVER used in this .h file. (a useless interface)
You'll need to draw class (struct/interface) diagrams of this .h file, its users, and its (internal) implementations and review the architecture. If you think the implementation deadline is too imminent for such architecture review & refactoring, you can postpone architecture review & refactoring after the M2 release.

Comment (gichan-jang) [reply include/nnstreamer-edge.h:81]: `nns_edge_flag_e` is used for creating the edge handle.
```
int nns_edge_create_handle (const char *id, nns_edge_connect_type_e connect_type, int flags, nns_edge_h *edge_h);
```

Comment (myungjoo) [reply include/nnstreamer-edge.h:81]: > `nns_edge_flag_e` is used for creating the edge handle.
> 
> ```
> int nns_edge_create_handle (const char *id, nns_edge_connect_type_e connect_type, int flags, nns_edge_h *edge_h);
> ```

Then, change ```int flags``` --> ```nns_edge_flag_e flags```.

Comment (gichan-jang) [reply include/nnstreamer-edge.h:81]: @anychhoice , Please update the flags and data type.

Comment (anychhoice) [reply include/nnstreamer-edge.h:81]:
```
typedef enum {
  NNS_EDGE_ROLE_QUERY_CLIENT = 0,
  NNS_EDGE_ROLE_QUERY_SERVER,
  NNS_EDGE_ROLE_PUB,
  NNS_EDGE_ROLE_SUB,

  NNS_EDGE_ROLE_UNKNOWN,
} nns_edge_role_e;
```

How about change name to `nns_edge_role_e` ?

Comment (gichan-jang) [reply include/nnstreamer-edge.h:81]: Nice suggestion. But I think `nns_edge_node_type_e` is more appropriate.

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c [hunk 3]
```diff
@@ -931,10 +876,12 @@ _nns_edge_send_thread (void *thread_data)
...
           conn_data = (nns_edge_conn_data_s *) eh->connections;
           while (conn_data) {
-            /** @todo update code for each connect type */
             conn = conn_data->sink_conn;
-            _nns_edge_transfer_data (conn, data_h, conn_data->id);
-
+            ret = _nns_edge_transfer_data (conn, data_h, conn_data->id);
+            if (NNS_EDGE_ERROR_NONE != ret) {
+              nns_edge_loge ("Failed to transfer data. Close the connection.");
+              _nns_edge_remove_connection_by_conn (eh, conn);
```
Comment (jaeyun-jung) [inline src/libnnstreamer-edge/nnstreamer-edge-internal.c:883]: Suggestion: 
1. _nns_edge_remove_connection_by_conn() is unnecessary.
2. Need fix assertion case (set next connection data before releasing it)

```
client_id = conn_data->id;
conn = conn_data->sink_conn;
ret = _nns_edge_transfer_data (conn, data_h, client_id);
conn_data = conn_data->next;

if (NNS_EDGE_ERROR_NONE != ret) {
  nns_edge_loge ("Failed to transfer data. Close the connection.");
   _nns_edge_remove_connection (eh, client_id);
}
```

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c [hunk 4]
```diff
@@ -1573,6 +1628,33 @@ nns_edge_disconnect (nns_edge_h edge_h)
   return NNS_EDGE_ERROR_NONE;
 }
 
+/**
+ * @brief Thread to send data.
```
Comment (jaeyun-jung) [inline src/libnnstreamer-edge/nnstreamer-edge-internal.c:1632]: Need update description

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c (hunk 3 above)
Comment (gichan-jang) [reply src/libnnstreamer-edge/nnstreamer-edge-internal.c:883]: Thanks! Updated.

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c (hunk 4 above)
Comment (gichan-jang) [reply src/libnnstreamer-edge/nnstreamer-edge-internal.c:1632]: Updated.

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**